### ISSUE TITLE
Add more eprintln!() messages for protocol problems

### ISSUE THREAD

**Issue by vberger** [Smithay/wayland-rs#347]
In particular, rust-impl `wayland-client` should `eprintln!()` the issues when it encounters a problem with what the server has sent.

**Comment by Stargateur** [comment]
Couldn't be better to use [`log`](https://crates.io/crates/log) or the quite new but very promising [`tracing`](https://crates.io/crates/tracing) ?

**Comment by vberger** [comment]
I'm indecisive about that. These messages are printed in result of a protocol error, meaning that the connection is going to be closed, and in 99% of cases meaning the client will `panic!()` because `event_queue.dispatch()` returns an error.

Also, such messages being triggered is 100% guaranteed to be because of a bug (either server-side or client-side), from my point of view they mostly would serve as additional context to a panic backtrace, so I'm not sure what would be actually gained from pulling in the `log` machinery, and requiring downstream apps to use it to see these messages.

**Comment by Stargateur** [comment]
My point was a little more general, unless I miss it, I don't see any logging strategy on this project.

While I don't like much log, tracing crate offer a very good way to have trace capability allowing a lot of debugging help while being able to completely be disable at compile time contrary to log. (Well if I also didn't miss anything).

While adding some `eprintln()` for error is nice, I think only application is expected to write error level log. Well, on the specific case you want, it's a bug on the lib side so yes I suppose it's make sense to use error level but the user should also print it as error (resulting to double print ?).

So this issue could become more general and add tracing (or log) that print a LOT more debug information instead of just limit it to error when finally the user should report the error as well.

Any bug report could add the log file with a trace level allowing us to have a better view of the problem. This would add a great value overtime both for user of this crate and dev of this crate.

A big con of tracing is that it's still new but I think it's totally usable even now, another problem is this add compile time.

Finish with my 2 cents

**Comment by vberger** [comment]
I see your point. The current logging strategy mimicks a lot that of the system `libwayland`, which is mostly:

- Print to stderr any unrecoverable error
- Print to stderr a dump of the protocol activity if `WAYLAND_DEBUG=1` is set
- Otherwise, be silent

Given a significant fraction of the uses of this crate have the `use_system_lib` feature flag enabled, we are not getting away from that. So currently the pure-rust backend mostly mimicks this strategy, trying to produce as useful error messages as possible.

Using a logging framework for the additional messages generated by `wayland-client` is definitely possible, especially for `trace`-level debugging for example.

I'm unsure about the discrepancies it might cause though. Some messages would be printed to `stderr` directly or to the logging framework depending on whether `use_system_lib` is enabled, which seems rather poor design in terms of consistency.

**Comment by vberger** [comment]
Although, with #381 coming and the clean split between backend & frontend of the library, it'd be possible to contain the "direct to stderr" strategy to the backend only, and use real logging in the frontent part. This way the consistency would be preserved.

**Comment by Stargateur** [comment]
It's hard to answer these questions:

1. I don't think this crate should copy the coding style of C implementation of wayland (for many reason :p), that of course my opinion, I believe your goal was to have a pure Rust implementation, so have better code than C, not to limit it to C feature. That should include logging strategy.
2. Handle the log config and so `WAYLAND_DEBUG` must be transfer to the user control (some helper function not a big deal I think) cause user in Rust have the logging control, the library have no control.
3. I don't think it's inconsistency to have a different log output if the user enable a compile time feature of a crate. Also, it's about logging so not a big deal in my opinion. As I say the logging configuration should be let to user, and so the user can handle it as wanted.

I understand the problems you raise so I think this would need a proof of concept PR so see what it's could look like.

Use case:

- Rust call wayland-rs: Can use any log config without problem.
- Something call wayland-rs: Problem cause Something need to init a logger somehow. Maybe using a feature ?
- Rust call wayland-rs using system-lib: potentially mixing log format output, could be handle by the user but meh. Already happen in a lot of other solution.

We could also have a feature that initialize the logging system (for foreign code) in the same place where the crate already check the `WAYLAND_DEBUG` but it's wrong https://github.com/Smithay/wayland-rs/blob/c080f7519aa9e0ec60051f3342f90289c7c77cda/wayland-client/src/rust_imp/display.rs#L30

We could use a feature that switch between the two logging strategy. That require more code.

We could not care.

**Comment by vberger** [comment]
Well, #381 is definitely the occasion to rethink all that can be rethought, so it's definitely worth considering.

For sure, documenting that the logging info will be subpar if the system library is used, and moving the protocol debugging logic to a `debug` loglevel of the crate is something that can be considered.

**Comment by vberger** [comment]
After some digging, it appears that the system wayland-libs do allow us to hook a function to redirect logging, so this is something we can work with to redirect error messages to rust's `log`.

The main limitation is that protocol events dumping with `WAYLAND_DEBUG` cannot be redirected though.

**Comment by Stargateur** [comment]
We could also keep both, implement logging that is close to `WAYLAND_DEBUG` and use tracing only for `RUST_LOG` that could make sense.

But this increase the maintenance cost.

**Comment by vberger** [comment]
Not sure I understood correctly, do you mean keep the direct dumping of protocol events to stderr when `WAYLAND_DEBUG` is set, and in parallel also dump it to `RUST_LOG` at a debug or trace loglevel?

**Comment by Stargateur** [comment]
exactly, that not that bad, that would make opt-in the mixing style of logging. User could control what they want to debug, wayland or rust.

**Comment by vberger** [comment]
Yeah, I guess that could work reasonably well indeed.

**Comment by vberger** [comment]
This is going to be the fun part: the system libraries require a function matching the following interface to handle the logging hijack: 

```c
typedef void (*wl_log_func_t)(const char *, va_list);
```

**Comment by Stargateur** [comment]
well, that was to be expected, https://github.com/rust-lang/rust/issues/44930

I guess write at the hand the c function that will call Rust will be more simple and portable than require nightly unstable feature of trying to handle the X different implementation of C variadic function.

**Comment by Stargateur** [comment]
```c
#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <error.h>

typedef void (wl_log_func_t)(const char *, va_list);

void wl_log_funt(char const *fmt, va_list list);
void wl_log_funt(char const *fmt, va_list list) {
    va_list cpy;
    va_copy(cpy, list);
    int expected_size = vsnprintf(NULL, 0, fmt, cpy);
    if (expected_size <= 0) {
        error(EXIT_FAILURE, expected_size, "Unexpected Error");
        return;
    }
    va_end(cpy);
    
    size_t buffer_size = (size_t)expected_size + 1;
    char *buffer = malloc(sizeof *buffer * buffer_size);
    if (buffer == NULL) {
        error(EXIT_FAILURE, 0, "Unexpected Error");
        return;
    }
    
    int result_size = vsnprintf(buffer, buffer_size, fmt, list);    
    if (result_size != expected_size) {
        free(buffer);
        error(EXIT_FAILURE, result_size, "Unexpected Error");
        return;
    }
    // call rust
    printf("%s", buffer);
    free(buffer);
    
    return;
}

static void wrap(wl_log_func_t *f, char const *fmt, ...) {
    va_list list;
    va_start(list, fmt);
    (*f)(fmt, list);
    va_end(list);
}

int main(void)
{
    wrap(wl_log_funt, "%s: %d", "Long time no C", 42);
}
```

The problem is that is allocate memory but I least it's work well.

**Comment by vberger** [comment]
This is implemented in 0.30
